Optimize `Seek::stream_len` impl for `File`

tbu- opened the pull request:
It uses the file metadata on Unix with a fallback for files incorrectly reported as zero-sized. It uses `GetFileSizeEx` on Windows.

This reduces the number of syscalls needed for determining the file size of an open file from 3 to 1.

tbu- committed:
Optimize `Seek::stream_len` impl for `File`

It uses the file metadata on Unix with a fallback for files incorrectly
reported as zero-sized. It uses `GetFileSizeEx` on Windows.

This reduces the number of syscalls needed for determining the file size
of an open file from 3 to 1.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9068050488/job/24914473120) [(plain)](https://github.com/rust-lang/rust/commit/42b9823911034fd2d396a5d6f3b2414a1599e002/checks/24914473120/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking std_detect v0.1.5 (/checkout/library/stdarch/crates/std_detect)
error[E0308]: mismatched types
   --> library/std/src/sys/pal/windows/fs.rs:506:14
    |
506 |           Some(cvt(unsafe { c::GetFileSizeEx(self.handle.as_raw_handle(), &mut result) })
    |  _________----_^
    | |         arguments to this enum variant are incorrect
    | |         arguments to this enum variant are incorrect
507 | |             .map(|_| result))
    | |____________________________^ expected `Result<u64, Error>`, found `Result<i64, Error>`
    = note: expected enum `core::result::Result<u64, _>`
               found enum `core::result::Result<i64, _>`
               found enum `core::result::Result<i64, _>`
help: the type constructed contains `core::result::Result<i64, io::error::Error>` due to the type of the argument passed
   --> library/std/src/sys/pal/windows/fs.rs:506:9
    |
506 |            Some(cvt(unsafe { c::GetFileSizeEx(self.handle.as_raw_handle(), &mut result) })
    |  __________^____-
    | | _________|
    | ||
507 | ||             .map(|_| result))
    | ||____________________________-^
    |                               this argument influences the type of `Some`
note: tuple variant defined here
   --> /checkout/library/core/src/option.rs:579:5
    |
```

</details>

ChrisDenton commented on library/std/src/io/mod.rs:
```diff
@@ -2066,6 +2057,21 @@ pub trait Seek {
 
+#[doc(hidden)]
+#[unstable(feature = "internal_stream_len_default", issue = "none")]
+pub fn stream_len_default<T: Seek + ?Sized>(self_: &mut T) -> Result<u64> {
```
Can't this just be `pub(crate)` rather than `doc(hidden)`?

tbu- replied:
Changed.

paolobarbolini commented on library/std/src/fs.rs:
```diff
@@ -872,6 +901,9 @@ impl Seek for File {
         (&*self).seek(pos)
     }
+    fn stream_len(&mut self) -> io::Result<u64> {
+        (&*self).stream_len()
```
You've missed the `Seek` implementation for `Arc<File>` below

tbu- replied:
Thanks for noticing. Added.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9148826719/job/25151788266) [(plain)](https://github.com/rust-lang/rust/commit/af59081151cfabf5538d0e9a3be3ff5201db6c74/checks/25151788266/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.8s done
#16 DONE 35.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
   Compiling addr2line v0.21.0
error: function cannot return without recursing
   --> library/std/src/fs.rs:953:5
    |
953 |     fn stream_len(&mut self) -> io::Result<u64> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
954 |         (&*self).stream_len()
    |         --------------------- recursive call site
    |
    = help: a `loop` may express intention better if this is on purpose
    = note: `-D unconditional-recursion` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(unconditional_recursion)]`

error[E0596]: cannot borrow data in a `&` reference as mutable
    |
    |
954 |         (&*self).stream_len()
    |         ^^^^^^^^ cannot borrow as mutable
For more information about this error, try `rustc --explain E0596`.
error: could not compile `std` (lib) due to 2 previous errors
Build completed unsuccessfully in 0:00:15
  local time: Sun May 19 15:52:39 UTC 2024
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #127360) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

the8472 commented:
>  with a fallback for files incorrectly reported as zero-sized. 

Alas, this is insufficient. There are other filesystems with questionable `stat` impls that return incorrect values that aren't 0. For example sysfs.

```
$ stat -c %s /sys/kernel/oops_count 
4096
$ wc -c /sys/kernel/oops_count 
2 /sys/kernel/oops_count
```

tbu- commented:
> There are other filesystems with questionable `stat` impls that return incorrect values that aren't 0. For example sysfs.

Amazing, I did not foresee this.

> Alas, this is insufficient.

For this particular example, this doesn't look insufficient, since it doesn't regress anything:

```rs
use std::fs::File;
use std::io;
use std::io::Seek as _;
use std::io::SeekFrom;

fn main() -> io::Result<()> {
    let file = File::open("/sys/kernel/oops_count")?;
    println!("{}", (&file).seek(SeekFrom::End(0))?);
    Ok(())
}
```

This outputs 4096 on my machine, the same as the `stat` result.

```
[…]
openat(AT_FDCWD, "/sys/kernel/oops_count", O_RDONLY|O_CLOEXEC) = 3
lseek(3, 0, SEEK_END)                   = 4096
write(1, "4096\n", 5)                   = 5
close(3)                                = 0
```

tbu- commented:
It seems that `wc` falls back to reading the entire file if the file size is a multiple of the page size (4096 on my machine). I don't think we want to do that in Rust, that sounds horribly inefficient for files that happen to be a multiple of 4096 in size.

https://github.com/coreutils/coreutils/blob/74ef0ac8a56b36ed3d0277c3876fefcbf434d0b6/src/wc.c#L357-L365

the8472 commented:
Ok, maybe seeking is nonsense on those files too and they're only meant to be used via `read` and `write`.

But that still leaves FUSE drives which can implement basically arbitrary behavior. `stat` and `seek` aren't required to be consistent there either and seek might be the method doing the right thing.

And any approach distinguishing good from bad filesystems would require additional syscalls.

tbu- commented:
> But that still leaves FUSE drivers which can implement basically arbitrary behavior. `stat` and `seek` aren't required to be consistent there either and seek might be the method doing the right thing.

What is Rust's platform support for bad FUSE drivers? Who decides which of the methods to trust (`stat` vs `seek`)?

I can kinda see not crashing the Rust program if a FUSE driver decides to return `EBADF` for `close`. But this seems like a case of garbage in, garbage out, and it's on the user to not use file systems that lie to programs.

the8472 commented:
FUSE is just the canary in the coal mine. sysfs shows that in-kernel filesystems do strange things too. If we have 2 examples already then I assume there's some network filesystem will also do weird things when weird remote machines are involved.

So I'd say the trust hierarchy when it comes to "how many bytes does this file contain is" read < seek < stat

It's unfortunate that `statx` has a return field that's supposed to indicate what is actually supported by the kernel, but they're not using it to signal that reporting the size is unsupported on sysfs.

tbu- commented:
> FUSE is just the canary in the coal mine. sysfs shows that in-kernel filesystems do strange things too. If we have 2 examples already then I assume there's some network filesystem will also do weird things when weird remote machines are involved.

I had assumed `lseek` would error out for procfs, but turns out it doesn't:

```
$ python -c 'import os; print(open("/proc/1/cmdline").seek(0, os.SEEK_END))'
0
```

So after checking again, it seems that even for procfs `lseek` offers no advantage over `fstat`. This means the two examples (procfs, sysfs) we found are not examples where the two methods behave differently.

ChrisDenton commented:
Sorry, just catching up. Given the above, @the8472 are you happy that this is at least no worse than the status quo? I.e. if seek or stat is returning wrong or nonsense values then that's a problem in either case.

the8472 commented:
I think it would be better to only do this for regular files. Non-regular ones have weird behavior in general. E.g. directories have a size according to `stat` but reading or seeking would error, so they don't have a meaningful stream length.

Even with that restriction I think someone will still eventually run into issues around FUSE, network filesystems or weird filesystems.

That said, stream_len is an unstable API and it's a convenience method, so we don't necessarily have to provide ironclad guarantees. Maybe we should just rephrase the documentation of `Seek::stream_len`. Currently it says 

> This method is implemented using up to three seek operations.

this would be better

> The default implementation uses up to three seek operations.


I mean, it always applies to trait methods that trait impls override them and can provide different behavior, so this isn't special. But perhaps it's better to not suggest that implementations will only ever behave exactly like that.

joshtriplett commented:
I think we should evaluate the proc / sysfs case separately from the FUSE case.

For the former, having a heuristic like "this is exactly 4096 bytes" seems...acceptable, if painful.

For the latter, I think if people write FUSE filesystems with suffciently weird behavior, it's fine to get issues like trusting the size. We shouldn't *crash*, but it's fine if we don't have ideal behavior.

ChrisDenton commented:
Given all the above I would be inclined to accept this PR, maybe with the adjustment to the documentation suggested by the8472. However, I'm not really comfortable unilaterally approving if not all libs-team members are on board.

Can I ask for a libs consensus here?

tbu- commented:
Summary:

This changes `File::stream_len` to use one syscall for getting the file's size instead of seeking back and forth.

It was initially believed that this might change the results, but we have not found any case in which that happens. **Both** the old and the new method return the same bogus results for /proc and /sys. We haven't found a case in which they differ.

A concern regarding FUSE filesystems was raised, which can return arbitrary data from `lseek` and `fstat` and the data might differ there if the FUSE filesystem has a bug.

joshtriplett commented:
@rfcbot merge

the8472 commented:
For the common case (regular files) this should make things more robust. No O_APPEND weirdness, no concurrency issues with updating the seek position. 👍

the8472 commented:
Can you also apply the wording change to the trait docs?  

> Maybe we should just rephrase the documentation of `Seek::stream_len`. Currently it says
> 
> > This method is implemented using up to three seek operations.
> 
> this would be better
> 
> > The default implementation uses up to three seek operations.

tbu- committed:
Clarify description of `Seek::stream_len`

It can only describe the inner workings of the default implementation,
other implementations might not be implemented using seeks at all.

tbu- commented:
Done.

bors commented:
:umbrella: The latest upstream changes (presumably #137065) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joshtriplett commented:
Apologies that this has waited so long after FCP to get merged.

r=me once rebased.

tbu- commented:
Rebased.

joshtriplett commented:
@bors r+ rollup

bors commented:
:pushpin: Commit a5299856060237676f744c9304f6167148103669 has been approved by `joshtriplett`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joshtriplett a5299856060237676f744c9304f6167148103669 -->
<!-- homu: {"type":"Approved","sha":"a5299856060237676f744c9304f6167148103669","approver":"joshtriplett","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r-
failed in https://github.com/rust-lang/rust/pull/137124#issuecomment-2661470902

tbu- commented:
Should be fixed.

bors commented:
:umbrella: The latest upstream changes (presumably #137176) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tbu- commented:
@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #138830) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tbu- commented on library/std/src/fs.rs:
```diff
@@ -1247,9 +1247,39 @@ impl Write for &File {
+    ///
+    /// [changes]: io#platform-specific-behavior
+    fn stream_len(&mut self) -> io::Result<u64> {
+        if let Some(result) = self.inner.size() {
```
Fixed.

tgross35 commented:
@bros try cancel

tgross35 commented:
^ So the actual tests passed, "bors build finished" just never got picked up by a runner

@bors r=ChrisDenton

bors commented:
:pushpin: Commit 4b1e28b21f78491d4ae187a52ebd5436edef6281 has been approved by `ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ChrisDenton 4b1e28b21f78491d4ae187a52ebd5436edef6281 -->
<!-- homu: {"type":"Approved","sha":"4b1e28b21f78491d4ae187a52ebd5436edef6281","approver":"ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
